wrappers/dbus.go

```diff
@@ -0,0 +1,188 @@
+	if err := t.Execute(&templateOut, serviceData); err != nil {
+		return nil, err
+	}
+	return templateOut.Bytes(), nil
+}
+
+// snapNameFromServiceFile returns the snap name for the D-Bus service activation file.
+func snapNameFromServiceFile(filename string) (owner string, err error) {
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return "", err
+	}
+	snapKey := []byte("\nX-Snap=")
+	pos := bytes.Index(content, snapKey)
+	if pos == -1 {
+		return "", nil
+	}
+	snapName := content[pos+len(snapKey):]
+	pos = bytes.IndexRune(snapName, '\n')
+	if pos != -1 {
+		snapName = snapName[:pos]
+	}
+	return string(snapName), nil
+}
+
+// snapServiceActivationFiles returns the list of service activation files for a snap.
+func snapServiceActivationFiles(dir, snapName string) (services []string, err error) {
+	glob := filepath.Join(dir, "*.service")
+	matches, err := filepath.Glob(glob)
+	if err != nil {
+		return nil, err
+	}
+	for _, match := range matches {
+		serviceSnap, err := snapNameFromServiceFile(match)
+		if err != nil {
+			return nil, err
+		}
+		if serviceSnap == snapName {
+			services = append(services, filepath.Base(match))
+		}
+	}
+	return services, nil
+}
+
+func AddSnapDBusActivationFiles(s *snap.Info) error {
+	if err := os.MkdirAll(dirs.SnapDBusSessionServicesDir, 0755); err != nil {
+		return err
+	}
+	if err := os.MkdirAll(dirs.SnapDBusSystemServicesDir, 0755); err != nil {
+		return err
+	}
+
+	// Make sure we include any service files that claim to have
+	// been written by the snap.
+	sessionServices, err := snapServiceActivationFiles(dirs.SnapDBusSessionServicesDir, s.InstanceName())
+	if err != nil {
+		return err
+	}
+	systemServices, err := snapServiceActivationFiles(dirs.SnapDBusSystemServicesDir, s.InstanceName())
+	if err != nil {
+		return err
+	}
+
+	sessionContent := make(map[string]osutil.FileState)
+	systemContent := make(map[string]osutil.FileState)
+
+	for _, app := range s.Apps {
+		if !app.IsService() {
+			continue
+		}
+
+		for _, slot := range app.ActivatesOn {
+			var busName string
+			if err := slot.Attr("name", &busName); err != nil {
+				return err
+			}
+
+			content, err := generateDBusActivationFile(app, busName)
+			if err != nil {
+				return err
+			}
+			filename := busName + ".service"
+			fileState := &osutil.MemoryFileState{
+				Content: content,
+				Mode:    0644,
+			}
+			switch app.DaemonScope {
+			case snap.SystemDaemon:
+				systemContent[filename] = fileState
+				systemServices = append(systemServices, filename)
+			case snap.UserDaemon:
+				sessionContent[filename] = fileState
+				sessionServices = append(sessionServices, filename)
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDBusSessionServicesDir, sessionServices, sessionContent)
+	if err != nil {
+		return err
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapDBusSystemServicesDir, systemServices, systemContent)
+	if err != nil {
+		// Clean up files installed by first invocation
```

Comment:
this will not put in place the previous state of session service files, it will remove old and new, I suppose it's what you want/expect?  but is not very clear

Reply:
I've changed the comment to make it clear it is removing files in the error case.